test.pl: watchdog: Simplify, clarify, recover from failure

khwilliamson opened the pull request:
@tonycoz noticed a simplification this makes.

If something goes wrong on WIndows and VMS, previously it gave up; now it tries an `alarm`.

It now doesn't try to use the thread method unless safe signals are in effect.

Some comments are clarified

* This set of changes does not require a perldelta entry.

khwilliamson committed:
test.pl: Clarify some comments, ternary, white space

Parenthesize the ternary conditional for clarity

Change some indentation, mostly in preparation for later commits

khwilliamson committed:
test.pl: Don't use watchdog thread unless safe signals

In the thread documentation, I noticed that unsafe signals preclude
killing threads with a signal.  But this is the normal method for
clearing a watchdog timer under threads.  I doubt that people are
compiling with PERL_OLD_SIGNALS these days, so I didn't check for that,
but it's easy enough to check for the environment variable that does the
same thing at runtime.

khwilliamson committed:
test.pl: Try alarm if can't start watchdog on Win/VMS

Prior to this commit, the code gave up immediately if it tried
unsuccessfully to use a watchdog thread.  But the alarm method is still
available.  Drop down to try it.

tonycoz commented on t/test.pl:
```diff
@@ -1952,8 +1952,7 @@ ($;$)
                 # process
-                eval("END { local \$! = 0; local \$? = 0;
-                            wait() if kill('KILL', $watchdog); };");
+                eval("END { watchdog(0); };");
```
Why eval? Why not just a top level end block?

khwilliamson replied:
I was just cargo culting without thinking much about it

tonycoz commented on t/test.pl:
```diff
@@ -1969,8 +1968,7 @@ ($;$)
                 # process
-                eval "END { local \$! = 0; local \$? = 0;
-                            wait() if kill('KILL', $watchdog); };";
+                END { watchdog(0); };
```
Like BEGIN, END is registered at compile-time, and since we only need to `watchdog(0)` once we only need one END block with watchdog(0) in it, not two.

This is why I suggested putting it at the top level -- it's not associated with any particular way to create the watchdog since watchdog(0) cancels any watchdog mechanism.

If you're worried about the cost (trivial) or side effects (run from potentially buggy code that we're testing here) you might keep a flag to indicate that an alarm() watchdog is set, but I don't think it's necessary.

khwilliamson replied:
It seemed wrong to run this on all test files when so few actually need it.  And I'm uncomfortable making it anything but a no-op when I change it to do that running.  So I added a flag to avoid an alarm(0).

khwilliamson committed:
test.pl: Change variable name

The next commit changes things, so that some people may find that after
that, the current name would be misleading or confusing.

tonycoz commented:
How is:
```
watchdog(5);
sleep 4;
watchdog(10);
sleep 8;
# eof
```
meant to behave?

Right now with the changes from this PR a simple test like:
```
#!perl
    BEGIN {
        chdir 't' if -d 't';
        require './test.pl';
        set_up_inc(qw '../lib ...');
    }

watchdog(5);
sleep 4;
watchdog(10);
sleep 8;
ok(1);
done_testing();
```
 I get:
```
$ ./perl t/try.t
# Test process timed out - terminating
Killed
```

One option here would be for all entries to the watchdog() function to clear existing watchdogs and then return immediately after if the timeout is zero

With that the above produces:
```
$ ./perl t/try.t
ok 1 - [at t/try.t line 12]
1..1
```

I tested with:
```
@@ -1849,7 +1849,6 @@ sub watchdog ($;$)
 
     # If cancelling, use the state variables to know which method was used to
     # create the watchdog.
-    if ($timeout == 0) {
 
         # Is a no-op if no watchdog is currently set
         if ($watchdog_thread) {
@@ -1865,8 +1864,8 @@ sub watchdog ($;$)
         undef $watchdog_thread;
         undef $watchdog_process;
         undef $watchdog_alarm;
-        return;
-    }
+        return if $timeout == 0;
+
```

khwilliamson committed:
test.pl: Don't assume existing watchdog uses alarm()

The fallback method of setting a watchdog timer is to use alarm().
Prior to this commit, the code presumed that if it wasn't one of the
other possibilities, it must be this one.  This is a valid assumption
currently, but future commits will change that.  Prepare for them with
this commit.

khwilliamson committed:
test.pl: Always set up (simpler) watchdog END block

This commit removes the existing END blocks that vary depending on the
type of timer being used, and replaces them with a single END block that
always exists, and simply calls watchdog(0).  This already cancels
whatever watchdogs are in place, doing nothing if none are.

Spotted by Tony Cook

Even though a test file should clear every timer before exit, this makes
sure it happens.

khwilliamson committed:
test.pl: Harden watchdog against glitches

Only one type of watchdog timer supposedly can exist.  But I can see the
possibility that a temporary glitch in the system caused a fallback type
to be used.  This commit simply makes sure that when cancelling,
anything out there gets cancelled, instead of assuming there's just one
possibility.

khwilliamson committed:
test.pl: Calling watchdog cancels existing ones

This enhances the code to work on cases like:

watchdog(5);
sleep 4;
watchdog(10);
sleep 8;

Prior to this commit, you had to explicitly cancel an existing watchdog
before setting a new one.

As a result of this commit, all possible wathdog variables get undef'd
early.  This makes later undef calls redundant, so they are removed.

Suggested by Tony Cook
